neo/Cryptography/ECC/ECDsa.cs

```diff
@@ -36,7 +36,7 @@ private BigInteger CalculateE(BigInteger n, byte[] message)
 
         public BigInteger[] GenerateSignature(byte[] message)
         {
-            if (privateKey == null) throw new InvalidOperationException();
+            if (privateKey == null || publicKey.IsInfinity) throw new InvalidOperationException();
```

Comment:
I'm not quite sure how it improves security. You only need a private key to sign something, so `publicKey` is not used in any way here.

I'd be a little more worried about `VerifySignature` where `publicKey` really is used.

Also, `public ECDsa(byte[] privateKey, ECCurve curve)` constructor could probably be improved in that you can derive a public key from private key and use the same `ECDsa` object both to sign and to verify things.

Reply:
> Also, public ECDsa(byte[] privateKey, ECCurve curve) constructor could probably be improved in that you can derive a public key from private key and use the same ECDsa object both to sign and to verify things.

NM, just noticed that it's being done already.

Reply:
And if you want to protect against bogus private key (from which a `publicKey` is derived that satisfies `publicKey.IsInfinity` condition) maybe it's better to do that in the constructor, seems more reliable to me.

Reply:
@roman-khimov Thanks for your suggestion, it has been fixed, plz have a review.

Reply:
I like the way constructors look now, but I don't think removing the `privateKey == null` check is correct. This class is a little strange in that it can be initialized both with private key and with public key, but objects derived from such initialization are not equivalent in their functionality: you obviously can do anything when you have a private key, but you can't sign anything with just a public key. So until it's possible to call `GenerateSignature()` without a private key in the object this check should stay.

Reply:
> but I don't think removing the privateKey == null check is correct. 

Done, I have reset the check.